### PR TITLE
variable check

### PR DESCRIPTION
**Pull request by christophehurpeau**
a first try to check variables

### CODE REVIEW

**Comment by christophehurpeau** [comment]
@gcanti Yes ! Great !

**Comment by christophehurpeau** [comment]
`const [a] = _assert(bar, _t.list(_t.String), '[a]');` using babel-generator

I added `Array.prototype.find` polyfill to fix travis, not sure that it is a good idea

Also, you should check `findTypeAnnotationInObjectPattern` and `findTypeAnnotationInArrayPattern` maybe we can do better

**Comment by gcanti** [comment]
Using `babel-generator` is a good idea.

> I added Array.prototype.find polyfill to fix travis, not sure that it is a good idea

Maybe we could just import `lodash.find`

```diff
+const typeAnnotation = find(objectTypeAnnotation.properties, propType => propType.key.name === property.key.name)
-const typeAnnotation = objectTypeAnnotation.properties.find(propType => propType.key.name === property.key.name)
```

File: src/index.js [hunk 1]
```diff
@@ -686,6 +707,78 @@ export default function ({ types: t, template }) {
...
+      const typeAnnotation = objectTypeAnnotation.properties.find(propType => propType.key.name === property.key.name)
+      if (!typeAnnotation) {
+        continue
+      }
+
+      if (t.isIdentifier(property.value) && name === property.value.name) {
+        return typeAnnotation.value
+      } else if (t.isObjectPattern(property.value)) {
+        const result = findTypeAnnotationInObjectPattern(name, property.value, typeAnnotation.value)
+        if (result) {
```
Comment (gcanti) [inline src/index.js]: is there a particular reason you don't just `return findTypeAnnotationInObjectPattern(name, property.value, typeAnnotation.value)` here?

File: src/index.js [hunk 2]
```diff
@@ -686,6 +707,78 @@ export default function ({ types: t, template }) {
...
+      element = arrayPattern.elements[i]
+      const typeAnnotation = isGenericArray ? arrayTypeAnnotation.typeParameters.params[0] : arrayTypeAnnotation.types[i]
+      if (!typeAnnotation) {
+        continue
+      }
+
+      if (t.isIdentifier(element)) {
+        return typeAnnotation
+      } else if (t.isObjectPattern(element)) {
+        const result = findTypeAnnotationInObjectPattern(name, element, typeAnnotation)
```
Comment (gcanti) [inline src/index.js]: same here

File: src/index.js [hunk 3]
```diff
@@ -842,6 +935,82 @@ export default function ({ types: t, template }) {
...
+        if (state.opts[SKIP_ASSERTS_OPTION]) {
+          return
+        }
+
+        const node = path.node
+        try {
+          node.declarations.forEach(declarator => {
+            const id = declarator.id
+
+            if (id.hasBeenTypeChecked) {
```
Comment (gcanti) [inline src/index.js]: not sure why you handle a `hasBeenTypeChecked` boolean, commenting it out doesn't affect the tests

File: src/index.js [hunk 4]
```diff
@@ -842,6 +935,82 @@ export default function ({ types: t, template }) {
...
+
+          let typeAnnotation
+          if (t.isIdentifier(node.left)) {
+            const name = node.left.name
+            const binding = scope.getBinding(name)
+            if (!binding || binding.path.type !== 'VariableDeclarator') {
+              return
+            }
+
+            if (binding.kind === 'const') {
```
Comment (gcanti) [inline src/index.js]: isn't this already handled by other babel plugins?

File: src/index.js [hunk 5]
```diff
@@ -842,6 +935,82 @@ export default function ({ types: t, template }) {
...
+            const typeAnnotation = id.typeAnnotation && id.typeAnnotation.typeAnnotation
+            id.savedTypeAnnotation = typeAnnotation
+
+            if (!declarator.init || !id.savedTypeAnnotation) {
+              return
+            }
+
+            declarator.init = getWrappedVariableDeclaratorInitWithTypeCheckAST(declarator)
+            id.hasBeenTypeChecked = true
+          })
```
Comment (gcanti) [inline src/index.js]: here and in `AssignmentExpression` a `hasAsserts = true` is missing (otherwise the `_assert` helper wont be inserted)

**Comment by gcanti** [comment]
Needs some polishing but overall looks good, there's a problem with type parameters though:

```js
class A<T> {
  method() {
    const x: T = 'a'
  }
}
```

compiles to

```js
class A {
  method() {
    const x = _assert('a', T, "x"); // will throw T is not defined at runtime
  }
}
```

Not sure if we want to support them in this first round.

File: src/index.js (hunk 1 above)
Comment (christophehurpeau) [inline src/index.js]: Because if it didn't find it, it has to continue the for of

File: src/index.js (hunk 3 above)
Comment (christophehurpeau) [inline src/index.js]: Yes I can remove it, it was for before when I was replacing path

File: src/index.js (hunk 1 above)
Comment (gcanti) [inline src/index.js]: opss.. completely missed the `for..of`, sorry

**Comment by gcanti** [comment]
Thanks @christophehurpeau, may I merge or you are still working on this?

**Comment by christophehurpeau** [comment]
Working on polymorphic classes, at least remove the assert to avoid breaking code

**Comment by gcanti** [comment]
I guess we can solve that as I did with functions:

```js
Class(path) {
  const node = path.node
  const typeParameters = getTypeParameters(node)
  path.traverse({
    'Function|VariableDeclaration|AssignmentExpression'(path) {
      const node = path.node
      // store type parameters so we can read them later
      node._typeParameters = assign(typeParameters, node._typeParameters)
    }
  })
}
```

and then pass `node._typeParameters` to `getWrappedVariableDeclaratorInitWithTypeCheckAST` and `getWrappedAssignmentWithTypeCheckAST`

EDIT: here's a test case

actual.js

```js
class A<T> {
  method() {
    let x: T = 'a'
    x = 'b'
  }
}
```

expected.js

```js
class A {
  method() {
    let x = _assert('a', _t.Any, 'x');
    x = _assert('b', _t.Any, 'x');
  }
}
```

**Comment by gcanti** [comment]
@christophehurpeau please write here or assign this PR to me when you are ok. I'll merge and then do some polishing (I already spotted a little defect in my previous code but I want to merge this before)

**Comment by christophehurpeau** [comment]
just added the support for [bounded polymorphism](https://flowtype.org/blog/2015/03/12/Bounded-Polymorphism.html), you can check it. I should probably have done it in other PR, I can split them if you want

**Comment by christophehurpeau** [comment]
Also removed useless assert when we know already the type of the value

**Comment by gcanti** [comment]
Wait @christophehurpeau, you are doing a great work, but yes, there are too many moving parts now. 

Please split your proposals in several PRs, so we can discuss them one by one. I suggest at least:

- add support for `const`s (`VariableDeclaration`, polymorphic classes included)
- add supoprt for `let`s (`AssignmentExpression`, polymorphic classes included)
- add support for bounded polymorphism (please open an issue with a proposal before working on this)

Also, please **do not modify existing tests** if not strictly required. Add new tests instead so we can be sure there are no regressions.

For what concerns your last commit, at first glance seems a feature I don't want to support, but let's talk about it. Again please open an issue with a proposal.

**Comment by christophehurpeau** [comment]
I don't understand, there is no difference between `const` and `let`

**Comment by gcanti** [comment]
The difference is that you want to support also assignement expressions other than variable declarations. If you like to implement them in one single shot, I'm fine with that. I'm just suggesting to keep the PRs as small as possible, so that tests and reviews are easier to check

**Comment by gcanti** [comment]
Closing in favour of https://github.com/gcanti/babel-plugin-tcomb/issues/70, thanks @christophehurpeau

**Comment by christophehurpeau** [comment]
You are closing a PR in favour of an issue ?

**Comment by gcanti** [comment]
@christophehurpeau in favour of a *discussion*. 

As stated above, this PR, while very fruitful (and I thank you again for your work on this), now contains too many mixed features, some of which I'd prefer to elaborate further.

**Pull request closed without merging**